contracts/marketplace/Marketplace.sol

```diff
@@ -103,85 +100,121 @@ contract Marketplace is Ownable, Pausable, Destructible {
-    address assetOwner = nonFungibleRegistry.ownerOf(assetId);
-    require(msg.sender == assetOwner);
-    require(nonFungibleRegistry.isAuthorized(address(this), assetId));
-    require(priceInWei > 0);
-    require(expiresAt > block.timestamp.add(1 minutes));
-
-    bytes32 auctionId = keccak256(
-      block.timestamp, 
-      assetOwner,
-      assetId, 
-      priceInWei
+  function createOrder(
+    uint256 assetId,
+    address nftAddress,
+    uint256 priceInWei,
+    uint256 expiresAt
+  )
+    public
+    whenNotPaused
+  {
+    require(isContract(nftAddress), "The NFT Address should be a contract");
+
+    ERC721 nftRegistry = ERC721(nftAddress);
+    address assetOwner = nftRegistry.ownerOf(assetId);
+
+    require(msg.sender == assetOwner, "Only the owner can create orders");
+    require(
+      nftRegistry.getApproved(assetId) == address(this) || nftRegistry.isApprovedForAll(assetOwner, address(this)),
+      "The contract is not authorized to manage the asset"
+    );
+    require(priceInWei > 0, "Price should be bigger than 0");
+    require(expiresAt > block.timestamp.add(1 minutes), "Expires should be bigger than 1 minute");
+
+    bytes32 orderId = keccak256(
+      abi.encodePacked(
+        block.timestamp,
+        assetOwner,
+        assetId,
+        priceInWei
+      )
     );
 
-    auctionByAssetId[assetId] = Auction({
-      id: auctionId,
+    orderByAssetId[nftAddress][assetId] = Order({
+      id: orderId,
       seller: assetOwner,
+      nftAddress: nftAddress,
       price: priceInWei,
       expiresAt: expiresAt
     });
 
     // Check if there's a publication fee and
-    // transfer the amount to marketplace owner.
+    // transfer the amount to marketplace owner
     if (publicationFeeInWei > 0) {
-      require(acceptedToken.transferFrom(
+      acceptedToken.transferFrom(
         msg.sender,
         owner,
         publicationFeeInWei
-      ));
+      );
     }
 
-    emit AuctionCreated(
-      auctionId,
-      assetId, 
+    emit OrderCreated(
+      orderId,
+      assetId,
       assetOwner,
-      priceInWei, 
+      nftAddress,
+      priceInWei,
       expiresAt
     );
   }
 
   /**
     * @dev Cancel an already published order
-    *  can only be canceled by seller or the contract owner.
+    *  can only be canceled by seller or the contract owner
+    * @param nftAddress - Address of the NFT registry
     * @param assetId - ID of the published NFT
     */
-  function cancelOrder(uint256 assetId) public whenNotPaused {
-    require(auctionByAssetId[assetId].seller == msg.sender || msg.sender == owner);
-
-    bytes32 auctionId = auctionByAssetId[assetId].id;
-    address auctionSeller = auctionByAssetId[assetId].seller;
-    delete auctionByAssetId[assetId];
-
-    emit AuctionCancelled(auctionId, assetId, auctionSeller);
+  function cancelOrder(address nftAddress, uint256 assetId) public whenNotPaused {
+    Order storage order = orderByAssetId[nftAddress][assetId];
```

Comment:
I think we should use `memory` instead of `storage` due we are not using the pointer to change the contract storage, but only getting properties as `id`, `seller`, etc

[Reference Link](https://medium.com/coinmonks/storage-vs-memory-in-solidity-8251847186fd)